src/differencer/jsonDiff.ts

```diff
@@ -0,0 +1,382 @@
+    msg = `${JSON_LD_KEY_GRAPH} property in ${objLoc} json must be an array of json nodes`;
+    logError(msg);
+    return;
+  }
+}
+function logError(msg: string): never {
+  if (msg == null) {
+    msg = "Unknown error in json diff";
+  }
+  ramlToolLogger.error(msg);
+  throw new Error(msg);
+}
+/**
+ * Parse JSON differences and generate an array of typed diff objects
+ * @param diffs Differences JSON
+ *
+ * @returns Array of NodeDiff objects
+ */
+function parseDiffs(diffs: object): Array<NodeDiff> {
+  const typedDiffs: Array<NodeDiff> = [];
+  if (diffs == null || _.isEmpty(diffs)) {
+    //no changes
+    ramlToolLogger.info("No differences found by the jsondiffPatch");
+    return typedDiffs;
+  }
+
+  ramlToolLogger.debug("Parsing differences generated by jsondiffPatch");
+  const graphDiffs = diffs[JSON_LD_KEY_GRAPH];
+  if (graphDiffs != null) {
+    //process graph changes
+    ramlToolLogger.debug("Parsing differences in JSON-LD graph node");
+    typedDiffs.push(...parseGraph(graphDiffs));
+  }
+  const contextDiffs = diffs[JSON_LD_KEY_CONTEXT];
+  if (contextDiffs != null) {
+    //process context changes
+    ramlToolLogger.debug("Parsing differences in the JSON-LD context node");
+    const typedContextDiff = parseNodePropDiffs(
+      JSON_LD_KEY_CONTEXT,
+      contextDiffs
+    );
+    if (typedContextDiff != null) {
+      typedDiffs.push(typedContextDiff);
+    }
+  }
+  return typedDiffs;
+}
+
+/**
+ * Parse differences for each JSON node in the graph
+ * @param graphDiffs Array of JSON nodes with differences
+ *
+ * @returns Array of NodeDiff objects
+ */
+function parseGraph(graphDiffs: Array<object>): Array<NodeDiff> {
+  const typedDiffs: Array<NodeDiff> = [];
+  Object.keys(graphDiffs).forEach(key => {
+    if (key === JSON_DIFF_PATCH_ARRAY_KEY) {
+      //Ignoring the property added by jsondiffpath to indicate that graph is an array
+      return;
+    }
+    const diff = graphDiffs[key];
+    let typedDiff;
+    if (_.isArray(diff)) {
+      //Node is added/removed/moved
+      typedDiff = addNodeDiff(diff);
+    } else if (_.isObject(diff)) {
+      //properties in a node are added/removed/modified/moved
+      typedDiff = parseNodePropDiffs(diff[JSON_LD_KEY_ID], diff);
+    }
+    if (typedDiff != null) {
+      typedDiffs.push(typedDiff);
+    }
+  });
+  return typedDiffs;
+}
+
+/**
+ * Parse differences to the node properties
+ * @param nodeId ID of the node
+ * @param diff JSON node that has differences
+ *
+ * @returns Instance of NodeDiff for the node if there are differences
+ */
+function parseNodePropDiffs(nodeId: string, diff: object): NodeDiff {
+  let typedDiff = new NodeDiff(nodeId);
+  Object.keys(diff)
+    .filter(key => key !== JSON_LD_KEY_ID) //ignore node id
+    .forEach(key => {
+      const value = diff[key];
+
+      //check if the property of the node is an array
+      const isArray =
+        value.hasOwnProperty(JSON_DIFF_PATCH_ARRAY_KEY) &&
```

Comment:
Would we ever expect that this property does _not_ exist on `value`, but _does_ exist on its prototype? If not, then I don't think that this check is necessary.